### ISSUE TITLE
Microsoft.Build.Exceptions.InvalidProjectFileException

### ISSUE THREAD

**Issue by KostadinMinov** [sensslen/nuget-license#96]
# Problem

I am using .NET Framework 4.7.2 in an old project and below is the error message I get. The error doesn't appear when I build the project - only when i run the nuget-license command against it. I also tried adding net472 as --target-framework argument but without success.

# "Solution"
I solved this manually through moving my VisualStudio folder to the directory the MSBuild is looking for as shown [here](https://stack247.wordpress.com/2011/08/29/the-imported-project-microsoft-webapplications-targets-was-not-found/). This isn't ideal as every other developer who has to run the project has to do the same.

The error:
```
Microsoft.Build.Exceptions.InvalidProjectFileException: The imported project "C:\Program Files\dotnet\sdk\8.0.403\Microsoft\VisualStudio\v17.0\WebApplications\Microsoft.WebApplication.targets" was not found. Confirm that the expression in the Import declaration "C:\Program Files\dotnet\sdk\8.0.403\Microsoft\VisualStudio\v17.0\WebApplications\Microsoft.WebApplication.targets" is correct, and that the file exists on disk. 
13>     at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject(String errorSubCategoryResourceName, IElementLocation elementLocation, String resourceName, Object[] args)
13>     at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject[T1,T2](IElementLocation elementLocation, String resourceName, T1 arg0, T2 arg1)
13>     at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImportsFromUnescapedImportExpression(String directoryOfImportingFile, ProjectImportElement importElement, String unescapedExpression, Boolean throwOnFileNotExistsError, List`1& imports)
13>     at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImportsFromUnescapedImportExpressionConditioned(String directoryOfImportingFile, ProjectImportElement importElement, List`1& projects, SdkResult& sdkResult)
13>     at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImports(String directoryOfImportingFile, ProjectImportElement importElement, SdkResult& sdkResult)
13>     at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
13>     at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
13>     at Microsoft.Build.Evaluation.Evaluator`4.Evaluate()
13>     at Microsoft.Build.Evaluation.Evaluator`4.Evaluate(IEvaluatorData`4 data, Project project, ProjectRootElement root, ProjectLoadSettings loadSettings, Int32 maxNodeCount, PropertyDictionary`1 environmentProperties, ILoggingService loggingService, IItemFactory`2 itemFactory, IToolsetProvider toolsetProvider, IDirectoryCacheFactory directoryCacheFactory, ProjectRootElementCacheBase projectRootElementCache, BuildEventContext buildEventContext, ISdkResolverService sdkResolverService, Int32 submissionId, EvaluationContext evaluationContext, Boolean interactive)
13>     at Microsoft.Build.Evaluation.Project.ProjectImpl.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
13>     at Microsoft.Build.Evaluation.Project.ProjectImpl.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
13>     at Microsoft.Build.Evaluation.Project.ProjectImpl.Initialize(IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext, Boolean interactive)
13>     at Microsoft.Build.Evaluation.Project..ctor(String projectFile, IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext, IDirectoryCacheFactory directoryCacheFactory, Boolean interactive)
13>     at Microsoft.Build.Evaluation.ProjectCollection.LoadProject(String fileName, IDictionary`2 globalProperties, String toolsVersion)
13>     at NuGetUtility.Wrapper.MsBuildWrapper.MsBuildAbstraction.GetProject(String projectPath) in D:\a\nuget-license\nuget-license\src\NuGetUtility\Wrapper\MsBuildWrapper\MsBuildAbstraction.cs:line 30
13>     at NuGetUtility.ReferencedPackagesReader.ReferencedPackageReader.GetInstalledPackages(String projectPath, Boolean includeTransitive, String targetFramework) in D:\a\nuget-license\nuget-license\src\NuGetUtility\ReferencedPackagesReader\ReferencedPackageReader.cs:line 30
13>     at NuGetUtility.Program.<>c__DisplayClass40_0.<OnExecuteAsync>b__1(String p) in D:\a\nuget-license\nuget-license\src\NuGetUtility\Program.cs:line 131
```

**Comment by sensslen** [comment]
nuget-license makes use of https://github.com/microsoft/MSBuildLocator. Would you mind reporting the issue there?

**Comment by KostadinMinov** [comment]
Thanks for your answer, after further investigation the issue was in the Microsoft.Web.WebJobs.Publish v2.0.0 package - it didn't like being built with MSBuild. 

Our team decided it would be better to skip the nuget-license command for the specific project that had errors building.

**Comment by gavinschultz** [comment]
@sensslen I did try that - forgot to mention it initially - I too was surprised it still had the same error. However I noticed that neither `VSToolsPath` nor `MSBuildExtensionsPath` are set in any .targets files I could find locally, nor are they set as environment variables from a developer command prompt. They _do_ seem to be a property defined by msbuild itself though - I see both VSToolsPath and MSBuildExtensionsPath [defined in the msbuild app.config](https://github.com/dotnet/msbuild/blob/428195cb74072bbb5314889119525369cb9bdeb0/src/MSBuild/app.config#L173-L191) file.

In any case I've determined that this error occurs only when I'm using the dotnet version of the tool (i.e. installed by `dotnet tool install`, or using `NuGetLicenseCore.exe` directly). Once I switch specifically to use `NuGetLicenseFramework.exe`, the problem is resolved.

Just hazarding a guess, I think likely stems from the behaviour of [the call to MSBuildLocator.RegisterDefaults()](https://github.com/sensslen/nuget-license/blob/v3.1.1/src/NuGetUtility/Wrapper/MsBuildWrapper/MsBuildAbstraction.cs#L38) which will set up (at least partially) the environment in which the project is loaded.

Digging in to MSBuildLocator, from the dotnet version of the tool it would, when finding the Visual Studio instance, end up calling [the NETCOREAPP code](https://github.com/microsoft/MSBuildLocator/blob/v1.7.8/src/MSBuildLocator/MSBuildLocator.cs#L377-L382) which yields a `VisualStudioInstance` with a `DiscoveryType` of `DotNetSdk`.

This will end up calling `ApplyDotNetSdkEnvironmentVariables` [in this code](https://github.com/microsoft/MSBuildLocator/blob/v1.7.8/src/MSBuildLocator/MSBuildLocator.cs#L132-L137):
```
if (instance.DiscoveryType == DiscoveryType.DotNetSdk)
{
    // The dotnet cli sets up these environment variables when msbuild is invoked via `dotnet`,
    // but we are using msbuild dlls directly and therefore need to mimic that.
    ApplyDotNetSdkEnvironmentVariables(instance.MSBuildPath);
}
```

... which [sets MSBuildExtensionsPath] folder(https://github.com/microsoft/MSBuildLocator/blob/v1.7.8/src/MSBuildLocator/MSBuildLocator.cs#L330), but sets it to the .NET SDK folder rather than an MSBuild folder, and sets it as an environment variable:
```
var variables = new Dictionary<string, string>
{
    [MSBUILD_EXE_PATH] = Path.Combine(dotNetSdkPath, "MSBuild.dll"),
    [MSBuildExtensionsPath] = dotNetSdkPath,
    [MSBuildSDKsPath] = Path.Combine(dotNetSdkPath, "Sdks")
};

foreach (var kvp in variables)
{
    Environment.SetEnvironmentVariable(kvp.Key, kvp.Value);
}
```

That does seem to correspond to the stack trace from @KostadinMinov (and mine too) where it's looking for the .targets file in a location starting at a .NET SDK path i.e. "C:\Program Files\dotnet\sdk\8.0.403\".

So in short, the dotnet version of nuget-license will be mostly, but not _entirely_, compatible with .NET Framework projects, due to those differences you'll get in the environment. In this case specifically `MSBuildExtensionsPath` will be wrong, but there may be more.

Again, haven't debugged this to know for sure, but it seems plausible.

I don't know if there's an easy fix for this since the symptoms are ultimately enacted by MSBuild.Locator. But it might be worthwhile in the readme.md recommending that .NET Framework projects use the .NET Framework version of the tool.

**Comment by sensslen** [comment]
Thanks for that detailed explanation - I had run into similar issues with native projects which is why the .net framework version exists in the first place....
